text/0000-safecoin-alternative-implementation/0000-safecoin-alternative-implementation.md

```diff
@@ -0,0 +1,238 @@
+
+First, let's define the form of these balances:
+
+```rust
+struct Coin {
+    units: u32,
+    parts: u32,
+}
+```
+
+The `units` field will represent whole safecoins, and since the defined upper limit of issuable safecoin is `2^32`, this need be no bigger than a `u32`.
+
+The `parts` field represents the number of `2^-32`-th parts of a single safecoin.  Since the maximum value of a `u32` is `(2^32)-1`, `parts` will always represent less than a single safecoin.
+
+To represent client-owned accounts for holding safecoin, the following will be used:
+
+```rust
+struct CoinAccount {
+    id: PublicSigningKey,
+    balance: Coin,
+}
+```
+
+As with other network addressable entities, the `id` will represent the account's address on the network.  It will be managed by the close group of Vaults for that address, using a new persona: CoinManager.  The ID is also an asymmetric public signing key.
+
+A single client can create many `CoinAccount`s if desired, and there is no need to associate any of these with its main MAID account.  For example, a user may want to run multiple farmer Vaults and allocate a different `CoinAccount` for each to receive the farming rewards.
+
+When a client creates a new `CoinAccount` on the network, the request need only specify the `id`, since the initial balance will always be 0.  It is up to the client to create and manage the key pairs of all `CoinAccount`s it owns.  These could be stored on the network as part of its encrypted MAID account, or managed by a standalone wallet application.
+
+To pay for network resources (e.g. putting data) a `PaymentAccount` must exist where the account's ID is the same as the MAID's public key:
+
+```rust
+struct PaymentAccount {
+    details: CoinAccount,
+    app_rules: BTreeMap<PublicSigningKey, AppRules>,
+}
+
+enum AppRules {
+    Unrestricted,
+    SpendLimit {
+        remaining: Coin,
+    }
+}
+```
+
+The details of the `PaymentAccount` will be explained below at "[Paying for Network Resources](#paying-for-network-resources)".  Until we reach that point, just consider this as a wrapped `CoinAccount` and when `CoinAccount` is mentioned it means either of these types.
+
+
+
+### Farming
+
+Each section of the network will be responsible for a proportion of the total issuable safecoin equal to the proportion of the network address space it manages.  For instance, at the start when the network has a single section covering the full address space, that section will be responsible for all `2^32` safecoin.  When the section splits, each half will be responsible for `2^31` safecoin.
+
+This means that for a section with Prefix length `n`, it will be responsible for `2^(32-n)` safecoin.
+
```

Comment:
We probably need to define the responsibility here. i.e. responsible for 
Managing the issuance of coins 
Managing recycling of coins
etc.

Reply:
Hmm, yeah - that's a tricky one to nail down.  I suppose I really mean "responsible for managing the amount of farmed coin from that address space by ensuring it doesn't exceed the allotted amount or drop below zero."  I think I could fit that in as the second sentence of the first paragraph of this section, if that makes sense?

Reply:
Yip, sounds good to me.